test/libyul/YulName.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <boost/test/unit_test.hpp>
+
+#include <libyul/Dialect.h>
+#include <libyul/YulName.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace solidity::yul::test
+{
+
+BOOST_AUTO_TEST_SUITE(YulName)
+
+BOOST_AUTO_TEST_CASE(repository_generate_labels_for_derived_types)
+{
+	Dialect dialect{};
+	YulNameRepository nameRepository;
+	auto const test1 = nameRepository.defineName("test1");
+	BOOST_CHECK(!nameRepository.isDerivedName(test1));
+	auto const test1_1 = nameRepository.deriveName(test1);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_1));
+	auto const test1_2 = nameRepository.deriveName(test1);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_2));
+	auto const test1_2_1 = nameRepository.deriveName(test1_2);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_2_1));
+	auto const test2 = nameRepository.defineName("test2_1");
```

Comment:
Is this not being simply `"test2"` intentional or a mistake?

```suggestion
	auto const test2 = nameRepository.defineName("test2");
```

Reply:
After having gone through the whole test, it seems intentional. But in that case the variable naming would probably be clearer with something like:
- `test2` -> `test2_1`
- `test2_1` -> `test2_1_derived`